File: app/src/androidTest/java/com/futsch1/medtimer/BasicUITest.java
package com.futsch1.medtimer;


import static androidx.test.espresso.Espresso.pressBack;
import static com.adevinta.android.barista.assertion.BaristaCheckedAssertions.assertChecked;
import static com.adevinta.android.barista.assertion.BaristaCheckedAssertions.assertUnchecked;
import static com.adevinta.android.barista.assertion.BaristaVisibilityAssertions.assertContains;
import static com.adevinta.android.barista.assertion.BaristaVisibilityAssertions.assertDisplayed;
import static com.adevinta.android.barista.assertion.BaristaVisibilityAssertions.assertNotContains;
import static com.adevinta.android.barista.interaction.BaristaClickInteractions.clickOn;
import static com.adevinta.android.barista.interaction.BaristaDialogInteractions.clickDialogPositiveButton;
import static com.adevinta.android.barista.interaction.BaristaEditTextInteractions.clearText;
import static com.adevinta.android.barista.interaction.BaristaEditTextInteractions.writeTo;
import static com.adevinta.android.barista.interaction.BaristaKeyboardInteractions.closeKeyboard;
import static com.adevinta.android.barista.interaction.BaristaListInteractions.clickListItem;
import static com.adevinta.android.barista.interaction.BaristaListInteractions.clickListItemChild;
import static com.adevinta.android.barista.interaction.BaristaMenuClickInteractions.openMenu;
import static com.futsch1.medtimer.AndroidTestHelper.MainMenu.MEDICINES;
import static com.futsch1.medtimer.AndroidTestHelper.MainMenu.OVERVIEW;
import static com.futsch1.medtimer.AndroidTestHelper.navigateTo;
import static com.futsch1.medtimer.AndroidTestHelper.setDate;
import static com.futsch1.medtimer.AndroidTestHelper.setValue;

import android.view.View;
import android.widget.TextView;

import org.junit.Test;

import java.time.LocalTime;
import java.util.Calendar;
import java.util.concurrent.atomic.AtomicReference;

public class BasicUITest extends BaseTestHelper {

    public static final String TEST_2 = "Test (2)";

    @Test
    //@AllowFlaky(attempts = 1)
    public void basicUITest() {
        AndroidTestHelper.createMedicine(" Test ");
        AndroidTestHelper.createReminder("1", LocalTime.of(18, 0));

        clickOn(R.id.openAdvancedSettings);

        clickOn(R.string.dosing_instructions);
        clickOn(R.string.sample_instructions);
        clickListItem(0);

        pressBack();
        pressBack();

        clickOn(R.id.openAdvancedSettings);
        assertContains(R.string.before_meal);
        pressBack();

        writeTo(R.id.editAmount, " 2 ");
        pressBack();

        clickListItem(R.id.medicineList, 0);
        assertDisplayed(R.id.editAmount, "2");
        pressBack();

        AndroidTestHelper.navigateTo(AndroidTestHelper.MainMenu.OVERVIEW);
        assertContains(TEST_2);

        AndroidTestHelper.navigateTo(AndroidTestHelper.MainMenu.MEDICINES);
        clickListItem(R.id.medicineList, 0);
        writeTo(R.id.editMedicineName, " Test2 ");
        pressBack();

        AndroidTestHelper.navigateTo(AndroidTestHelper.MainMenu.OVERVIEW);
        assertContains("Test2 (2)");
    }

    @Test
    //@AllowFlaky(attempts = 1)
    public void menuHandlingTest() {
        AndroidTestHelper.createMedicine("Test");
        AndroidTestHelper.createReminder("1", LocalTime.of(12, 0));

        clickOn(R.id.openAdvancedSettings);

        Calendar cycleStart = Calendar.getInstance();
        cycleStart.set(2025, 1, 1);
        String cycleStartString = AndroidTestHelper.dateToString(cycleStart.getTime());
        clickOn(R.string.cycle_reminders);
        clickOn(R.string.cycle_start_date);
        setDate(cycleStart.getTime());
        clickOn(R.string.cycle_consecutive_days);
        setValue("5");
        clickOn(R.string.cycle_pause_days);
        setValue("6");
        pressBack();

        clickOn(R.string.remind_on_weekdays);
        clickOn(R.string.monday);
        clickOn(R.string.tuesday);
        clickDialogPositiveButton();

        clickOn(R.string.remind_on_days_of_month);
        clickOn("1");
        clickOn("3");
        clickDialogPositiveButton();

        pressBack();

        clickOn(R.id.openAdvancedSettings);

        clickOn(R.string.cycle_reminders);
        assertContains(cycleStartString);
        assertContains("5");
        assertContains("6");
        pressBack();

        clickOn(R.string.remind_on_weekdays);
        assertUnchecked(R.string.monday);
        assertUnchecked(R.string.tuesday);
        assertChecked(R.string.wednesday);
        clickDialogPositiveButton();

        clickOn(R.string.remind_on_days_of_month);
        assertChecked("1");
        assertUnchecked("2");
        assertChecked("3");
        clickDialogPositiveButton();
    }

    @Test
    //@AllowFlaky(attempts = 1)
    public void notesTest() {
        AndroidTestHelper.createMedicine("Test");

        // Test saving notes
        String notes = "Contains catnip\n\nmeow :3";

        clickOn(R.id.openNotes);
        writeTo(R.id.notes, notes);
        closeKeyboard();
        clickOn(R.id.confirmSaveNotes);

        // Check if the note is saved
        clickOn(R.id.openNotes);
        assertDisplayed(R.id.notes, notes);

        // Test cancelling saving notes
        clearText(R.id.notes);
        closeKeyboard();
        clickOn(R.id.cancelSaveNotes);

        // Check that the note is unmodified
        clickOn(R.id.openNotes);
        assertDisplayed(R.id.notes, notes);
    }

    @Test
    //@AllowFlaky(attempts = 1)
    public void appIntro() {
        openMenu();

        clickOn(R.string.show_intro);

        assertDisplayed(R.string.intro_welcome);
        assertDisplayed(R.string.intro_welcome_description);

        clickOn(com.github.appintro.R.id.skip);

        assertDisplayed(R.string.tab_overview);
    }

    @Test
    //@AllowFlaky(attempts = 1)
    public void overviewFilters() {
        AndroidTestHelper.createMedicine("Test");
        AndroidTestHelper.createIntervalReminder("2", 1000);

        pressBack();

        AndroidTestHelper.navigateTo(AndroidTestHelper.MainMenu.OVERVIEW);
        assertContains(TEST_2);

        clickOn(R.id.filterRaised);
        assertContains(TEST_2);
        clickOn(R.id.filterRaised);

        assertContains(TEST_2);

        clickOn(R.id.filterTaken);
        assertNotContains(TEST_2);
        clickOn(R.id.filterTaken);

        assertContains(TEST_2);

        clickListItemChild(R.id.reminders, 0, R.id.stateButton);
        clickOn(R.id.takenButton);

        clickOn(R.id.filterTaken);
        assertContains(TEST_2);
        clickOn(R.id.filterTaken);

        assertContains(TEST_2);

        clickOn(R.id.filterRaised);
        assertNotContains(TEST_2);
        clickOn(R.id.filterRaised);

        clickOn(R.id.filterSkipped);
        assertNotContains(TEST_2);
        clickOn(R.id.filterSkipped);

        clickOn(R.id.filterScheduled);
        assertNotContains(TEST_2);
        clickOn(R.id.filterScheduled);

        navigateTo(MEDICINES);
        clickOn("Test");
        AndroidTestHelper.createReminder("1", LocalTime.of(20, 0));

        navigateTo(OVERVIEW);

        assertContains("Test (1)");

        clickOn(R.id.filterScheduled);
        assertContains("Test (1)");
    }

    @Test
    //@AllowFlaky(attempts = 1)
    @SuppressWarnings("java:S2699") // Using internal assert
    public void overviewDaySelection() {
        clickOn("2");

        navigateTo(MEDICINES);

        clickOn(R.id.overviewFragment);

        AtomicReference<View> view = new AtomicReference<>();
        view.set(baristaRule.getActivityTestRule().getActivity().findViewById(R.id.overviewWeek));

        TextView currentDay = view.get().findViewWithTag("selected");
        internalAssert(currentDay.getText().equals("2"));

        navigateTo(OVERVIEW);
        view.set(baristaRule.getActivityTestRule().getActivity().findViewById(R.id.overviewWeek));
        currentDay = view.get().findViewWithTag("selected");
        internalAssert(currentDay.getText().equals("1"));
    }
}

File: app/src/androidTest/java/com/futsch1/medtimer/BasicUITest.kt
package com.futsch1.medtimer

import com.adevinta.android.barista.rule.flaky.AllowFlaky
import com.futsch1.medtimer.core.domain.model.OverviewFilter
import com.futsch1.medtimer.core.ui.R
import dagger.hilt.android.testing.HiltAndroidTest
import org.junit.Test
import java.text.DateFormat
import java.time.LocalDate
import java.util.Calendar
import kotlin.test.assertTrue
import kotlin.time.Duration.Companion.hours


@HiltAndroidTest
class BasicUITest : MedTimerTestBase() {
    @Test
    @AllowFlaky(attempts = 3)
    fun basicUITest() {
        medicines.create(" Test ")
        medicineEditor.addReminder("1", laterToday())

        reminders.inSettingsOf(0) {
            inDosingInstructions {
                useSampleInstruction(R.string.before_meal)
                assertDosingInstructions(getString(R.string.before_meal))
            }
        }

        reminders.inSettingsOf(0) {
            assertDosingInstructions(getString(R.string.before_meal))
        }

        medicineEditor.setAmount(" 2 ")

        medicines.clickItem(0)
        medicineEditor.assertAmount("2")
        reminders.duplicate(0)

        reminders.assertCount(2)

        navigation.toOverview()
        overview.assertEventContains(TEST_2)

        navigation.toMedicines()
        medicines.clickItem(0)
        medicineEditor.rename(" Test2 ")

        navigation.toOverview()
        overview.assertEventContains("Test2 (2)")
    }

    @Test
    @AllowFlaky(attempts = 3)
    fun menuHandlingTest() {
        seed.medicine("Test") { reminder("1", laterToday()) }

        medicines.clickItem(0)

        val cycleStart = Calendar.getInstance()
        cycleStart.set(2025, 1, 1)
        val cycleStartString =
            DateFormat.getDateInstance(DateFormat.SHORT).format(cycleStart.getTime())

        reminders.inSettingsOf(0) {
            inCycle {
                setCycleStartDate(cycleStart.time)
                setConsecutiveDays(5)
                setPauseDays(6)
            }

            inWeekdays {
                clickItem(R.string.monday)
                clickItem(R.string.tuesday)
            }

            inDaysOfMonth {
                clickItem("1")
                clickItem("3")
            }
        }

        reminders.inSettingsOf(0) {
            inCycle {
                assertCycleStartDate(cycleStartString)
                assertConsecutiveDays("5")
                assertPauseDays("6")
            }

            inWeekdays {
                assertItemNotChecked(R.string.monday)
                assertItemNotChecked(R.string.tuesday)
                assertItemChecked(R.string.wednesday)
            }

            inDaysOfMonth {
                assertItemChecked("1")
                assertItemNotChecked("2")
                assertItemChecked("3")
            }
        }
    }

    @Test
    @AllowFlaky(attempts = 3)
    fun notesTest() {
        seed.medicine("Test")

        medicines.clickItem(0)

        val notesText = "Contains catnip\n\nmeow :3"

        notes.save(notesText)

        medicines.clickItem(0)
        notes.assertContains(notesText)

        notes.clearAndDiscard()
        notes.assertContains(notesText)
    }

    @Test
    @AllowFlaky(attempts = 3)
    fun appIntro() {
        appIntro.show()
        appIntro.assertPage(R.string.intro_welcome, R.string.intro_welcome_description)
        appIntro.skip()

        navigation.assertOverviewShown()
    }

    @Test
    @AllowFlaky(attempts = 3)
    fun overviewFilters() {
        val medicineId = seed.medicine("Test") { intervalReminder("2", 24.hours) }

        navigation.toOverview()
        overview.assertEventContains(TEST_2)

        overview.toggleFilter(OverviewFilter.RAISED)
        overview.assertEventContains(TEST_2)
        overview.toggleFilter(OverviewFilter.RAISED)

        overview.assertEventContains(TEST_2)

        overview.toggleFilter(OverviewFilter.TAKEN)
        overview.assertEventCount(0)
        overview.toggleFilter(OverviewFilter.TAKEN)

        overview.assertEventContains(TEST_2)

        overview.clickEventState(0)
        overview.clickAction(R.string.taken)

        overview.toggleFilter(OverviewFilter.TAKEN)
        overview.assertEventContains(TEST_2)
        overview.toggleFilter(OverviewFilter.TAKEN)

        overview.assertEventContains(TEST_2)

        overview.toggleFilter(OverviewFilter.RAISED)
        overview.assertEventCount(0)
        overview.toggleFilter(OverviewFilter.RAISED)

        overview.toggleFilter(OverviewFilter.SKIPPED)
        overview.assertEventCount(0)
        overview.toggleFilter(OverviewFilter.SKIPPED)

        overview.toggleFilter(OverviewFilter.SCHEDULED)
        overview.assertEventCount(0)
        overview.toggleFilter(OverviewFilter.SCHEDULED)

        seed.remindersOf(medicineId) { reminder("1", laterToday()) }

        navigation.toOverview()

        overview.assertEventContains("Test (1)")

        overview.toggleFilter(OverviewFilter.SCHEDULED)
        overview.assertEventContains("Test (1)")
    }

    @Test
    @AllowFlaky(attempts = 3)
    fun overviewDaySelection() {
        menus.clickAppOption(R.string.generate_test_data_and_events)

        val today = LocalDate.now()
        val secondDay = today.plusDays(1)

        overview.selectDay(secondDay)
        assertTrue(overview.eventCount() > 0)

        navigation.toOverview()
        overview.assertDaySelected(today)

        navigation.toMedicines()
        navigation.toOverview()
        overview.assertDaySelected(today)
        assertTrue(overview.eventCount() > 0)

        overview.nextWeek()
        assertTrue(overview.eventCount() > 0)

        overview.previousWeek()
        overview.assertDaySelected(today)
        assertTrue(overview.eventCount() > 0)

        overview.previousWeek()
        assertTrue(overview.eventCount() > 0)
    }

    companion object {
        const val TEST_2: String = "Test (2)"
    }
}
